coreneuron/nrniv/nrn_acc_manager.cpp

```diff
@@ -42,371 +43,465 @@ void setup_nrnthreads_on_device(NrnThread* threads, int nthreads) {
 
     /* @todo: why dt is not setup at this moment? */
     for (i = 0; i < nthreads; i++) {
-        (threads + i)->_dt = dt;
+        NrnThread* nt = threads + i;
         /* this thread will be computed on GPU */
-        (threads + i)->compute_gpu = 1;
+        nt->compute_gpu = 1;
+        if (nt->end <= 0) {
+            // this is an empty thread and could be only artificial
+            // cells. In this case nothing is executed on gpu.
+            nt->compute_gpu = 0;
+        }
     }
 
-    /* -- copy NrnThread to device. this needs to be contigious vector because offset is used to
-     * find
-     * corresponding NrnThread using Point_process in NET_RECEIVE block
-     */
-    d_threads = (NrnThread*)acc_copyin(threads, sizeof(NrnThread) * nthreads);
-
-    if (interleave_info == NULL) {
-        printf("\n Warning: No permutation data? Required for linear algebra!");
+    if (nrn_ion_global_map_size) {
+        double** d_data =
+            (double**)acc_copyin(nrn_ion_global_map, sizeof(double*) * nrn_ion_global_map_size);
+        for (int j = 0; j < nrn_ion_global_map_size; j++) {
+            if (nrn_ion_global_map[j]) {
+                /* @todo: fix this constant size 3 :( */
+                double* d_mechmap = (double*)acc_copyin(nrn_ion_global_map[j], 3 * sizeof(double));
+                acc_memcpy_to_device(&(d_data[j]), &d_mechmap, sizeof(double*));
+            }
+        }
```

Comment:
this is duplicated for unified  and non-unified version. Could you make this as separate function to avoid code duplication?